### PR TITLE
inabox: add window.AMP.inaboxUnregisterIframe(iframe)

### PR DESCRIPTION
**Pull request by jeffkaufman**
The current API for inabox is that when you want to remove a frame you delete it
from ampInaboxIframes.  This means there isn't a way for inabox to run any
additional cleanup on frame removal, and currently there's a memory leak where
references to iframes (or their descendents) are maintained indefinitely in
InaboxMessagingHost.iframeMap_ and in the position observation listeners.

This changes the API from 'delete from a list of frames' to 'call a function',
which allows the function to do arbitrary cleanup like removing references.

I'm not that happy with the test, since it runs in a very mocked environment.
To test this manually I've added examples/inabox.host.memory.html which
repeatedly adds an iframe and removes it again.  This initially showed us as
leaking about 4MB of memory per refresh (desktop chrome; silly fake ad) but now
doesn't show any leaking.

### CODE REVIEW

File: ads/inabox/inabox-messaging-host.js [hunk 1]
```diff
@@ -139,10 +139,13 @@ export class InaboxMessagingHost {
       return true;
     }
 
-    this.registeredIframeSentinels_[request.sentinel] = true;
-    this.positionObserver_.observe(iframe, data => {
-      this.sendPosition_(request, source, origin, data);
-    });
+    const positionObserverHandle = this.positionObserver_.observe(
```
Comment (keithwrightbos) [inline ads/inabox/inabox-messaging-host.js]: Can we rename this to observeUnregisterFn?

File: ads/inabox/position-observer.js [hunk 2]
```diff
@@ -68,9 +72,20 @@ export class PositionObserver {
     }
     // Send the 1st ping immediately
     callback(this.getPositionEntry_(element));
-    this.positionObservable_.add(() => {
+    const positionObservationHandler = (() => {
       callback(this.getPositionEntry_(element));
     });
+    this.positionObservable_.add(positionObservationHandler);
```
Comment (keithwrightbos) [inline ads/inabox/position-observer.js]: Add actually returns UnlistenDef so would be cleaner to keep a reference to that than the handler

File: ads/inabox/inabox-messaging-host.js (hunk 1 above)
Comment (jeffkaufman) [reply ads/inabox/inabox-messaging-host.js]: done

File: ads/inabox/position-observer.js (hunk 2 above)
Comment (jeffkaufman) [reply ads/inabox/position-observer.js]: done; that's much nicer

**Comment by jeffkaufman** [comment]
> Can you add type definitions to the fields within InaboxMessagingHost constructor?

Done.

> Also, consider storing the position observer unlisten function in iframeMap_ as well that way all state to be cleaned is in one spot.

Actually, keeping an iframe -> cleanup work mapping would really be a lot nicer.  Then anything that does some work which will need to be cleaned up can add it to the map at that time.

(But I don't have time to do that before I leave for a week.)

**Comment by keithwrightbos** [comment]
Cloned into #13795 as Jeff is OOTO.

**Pull request closed without merging**